src/core/harpoon.ml

```diff
@@ -304,14 +304,130 @@ module Tactic (T : TacticContext) = struct
     |> solve' g
 end
 
+module Automation = struct
+  open Comp
+
+  type t = unit proof_state -> Tactic.tactic_context -> bool
+
+  let auto_intros : t =
+    fun g tctx ->
+    let (t, _) = g.goal in
+    dprintf
+      (fun p ->
+        p.fmt
+          "[auto_intros]: invoked on %a"
+          (P.fmt_ppr_cmp_typ g.context.cD Pretty.std_lvl) t
+      );
+    match t with
+    | TypArr _ | TypPiBox _ ->
+       Tactic.intros None g tctx;
+       true
+    | _ -> false
+
+  (** Solve {v ... -> P -> ... -> P v} case automatically.
+      For example,
+      this function will resolve
+
+        x: [|- a]
+      -------------
+      [|- a]
+   *)
+  let auto_solve_trivial : t =
+    fun g tctx ->
+    let { cD; cG; _ } = g.context in
+    let m_is_witness (m : LF.ctyp_decl) =
+      dprintf
+        (fun p ->
+          p.fmt
+            "@[<v>[auto_solve_trivial] witness candidate = %a@]"
+            (P.fmt_ppr_lf_ctyp_decl cD Pretty.std_lvl) m
+        );
+      match m with
+      | LF.Decl (_, mtyp, _) ->
+         Whnf.convCTyp g.goal (Comp.TypBox (Loc.ghost, mtyp), LF.MShift 0)
+      | LF.DeclOpt _ ->
+         raise (Error.Violation "[auto_solve_trivial] Unexpected DeclOpt")
+    in
+    let build_mwitness (m : LF.ctyp_decl * int) =
+      match m with
+      | (LF.Decl (_, LF.ClTyp (_, dctx), _), idx) ->
+         let open LF in
+         let open Loc in
+         let head = MVar (Offset idx, S.LF.id) in
+         let clobj = MObj (Root (ghost, head, Nil)) in
+         let psi_hat = Context.dctxToHat dctx in
+         Box (ghost, (ghost, ClObj (psi_hat, clobj)))
+      (** The following case is impossible because m_is_witness
+          will never return true for a DeclOpt.
+       *)
+      | _ ->
+         raise (Error.Violation "[auto_solve_trivial] Impossible case")
+    in
+    let c_is_witness (c : Comp.ctyp_decl) =
+      dprintf
+        (fun p ->
+          p.fmt
+            "@[<v>[auto_solve_trivial] witness candidate = %a@]"
+            (P.fmt_ppr_cmp_ctyp_decl cD Pretty.std_lvl) c
+        );
+      match c with
+      | Comp.CTypDecl (_, typ, _) ->
+         Whnf.convCTyp g.goal (typ, LF.MShift 0)
+      | Comp.CTypDeclOpt _ ->
+         raise (Error.Violation "[auto_solve_trivial] Unexpected CTypDeclOpt")
+      | Comp.WfRec _ ->
+         raise (Error.Violation "[auto_solve_trivial] Unexpected WfRec")
+    in
+    let build_cwitness (c : Comp.ctyp_decl * int) =
+      match c with
+      | (_, idx) ->
+         let open Comp in
+         let open Loc in
+         Syn (ghost, Var (ghost, idx))
+    in
+    let open Maybe in
+    let opt_mwitness =
+      lazy
+        (Context.find_with_index' cD m_is_witness
+         $> build_mwitness
+        )
+    in
+    let opt_cwitness =
+      lazy
+        (Context.find_with_index' cG c_is_witness
+         $> build_cwitness
+        )
+    in
+    let opt_witness = opt_mwitness <|> opt_cwitness in
+    match opt_witness with
+    | lazy None ->
```

Comment:
I wasn't aware you could use `lazy` in patterns! I would always call `Lazy.force` on the scrutinee before matching. This is excellent; thanks for showing me.